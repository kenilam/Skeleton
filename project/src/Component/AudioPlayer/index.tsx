import React, {DependencyList, useEffect, useState} from 'react';
import IAudio from './spec';

const AudioPlayer = ({url}: IAudio.Props) => {
  const [index]: IAudio.IndexState = useState(0);
  const [tracks, updateTracks]: IAudio.TracksState = useState();
  
  const controller = new AbortController();
  const {signal} = controller;
  
  useEffect(
    () => {
      if (tracks) {
        return;
      }
      
      window.fetch(url, {signal})
      .then(
        response => response.json()
      )
      .then(
        tracks => {
          updateTracks(tracks);
        }
      );
      
      return () => {
        controller.abort();
      };
    },
    [tracks] as DependencyList
  );
  
  return (
    <video controls={false} autoPlay={true}>
      <source src={`${process.env.API_URL}${tracks ? tracks[index].url : null}`} type='audio/mpeg' />
    </video>
  );
};

export default AudioPlayer;
