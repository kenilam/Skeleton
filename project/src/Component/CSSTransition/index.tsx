import classnames from 'classnames';
import React from 'react';
import {CSSTransition as Origin} from 'react-transition-group';
import {EnterHandler, ExitHandler} from 'react-transition-group/Transition';
import ICSSTransition from './spec';
import Style, {TransitionStyle, TransitionStyleName} from './Style';
import {addEndListener, classNameModifier} from './Utility';

const CSSTransition: React.FunctionComponent<ICSSTransition.Props> = (
  {
    appear = true,
    classNames,
    children,
    mountOnEnter = true,
    onEnter,
    onEntered,
    onEntering,
    onExit,
    onExited,
    onExiting,
    timeout = null,
    transitionIn,
    transitionKey,
    transitionStyle = 'custom',
    unmountOnExit = true
  }
) => {
  const onEnterHandler: EnterHandler = (node, isAppearing) => {
    classNameModifier.addDefault(node, TransitionStyle[transitionStyle]);
    onEnter && onEnter(node, isAppearing);
  };
  
  const onEnteredHandler: EnterHandler = (node, isAppearing) => {
    classNameModifier.removeDone(node, TransitionStyle[transitionStyle]);
    onEntered && onEntered(node, isAppearing);
  };
  
  const onExitHandler: ExitHandler = (node) => {
    classNameModifier.addDefault(node, TransitionStyle[transitionStyle]);
    onExit && onExit(node);
  };
  
  return (
    <Origin
      addEndListener={!timeout && addEndListener}
      appear={appear}
      classNames={classnames(classNames, Style.cssTransition)}
      in={transitionIn}
      key={transitionKey}
      mountOnEnter={mountOnEnter}
      onEnter={onEnterHandler}
      onEntered={onEnteredHandler}
      onEntering={onEntering}
      onExit={onExitHandler}
      onExited={onExited}
      onExiting={onExiting}
      timeout={timeout}
      unmountOnExit={unmountOnExit}
    >
      {children}
    </Origin>
  );
};

export {Style, TransitionStyle, TransitionStyleName};
export default CSSTransition;
