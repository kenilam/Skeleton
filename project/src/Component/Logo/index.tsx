import resources from '$/resources';
import {Link} from '@/Component';
import React from 'react';
import ILogo from './spec';
import Style from './Style';

const {
  siteName,
  view: {
    home: {path}
  }
} = resources;

const Logo: React.FunctionComponent<ILogo.Props> = () => (
  <h1 data-component={Style.default}>
    <Link
      to={path}
    >
      {siteName}
    </Link>
  </h1>
);

export default Logo;
