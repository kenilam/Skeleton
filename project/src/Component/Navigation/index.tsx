import resources from '$/resources';
import {Link} from '@/Component';
import INavigation from '@/Component/Navigation/spec';
import {RandomId} from '@/Helper';
import classnames from 'classnames';
import React from 'react';
import Style from './Style';

const {view} = resources;

const DEFAULT_ITEMS: INavigation.Links = Object.keys(view).map(
  (viewName) => {
    const {name, path} = view[viewName];
    
    return {
      children: name,
      to: path
    };
  }
).filter(({to}) => Boolean(to) && to !== view.home.path);

const Navigation: React.FunctionComponent<INavigation.Props> = ({className, inline = false, items, onClick}) => (
  <nav
    className={classnames(
      className,
      {
        'is-inline': inline
      }
    )}
    data-component={Style.default}
    role='navigation'
  >
    {
      (items || DEFAULT_ITEMS).map(
        ({children, to}) => (
          <Link
            to={to}
            onClick={onClick && (event => {
              event.preventDefault();
              onClick(event, to);
            })}
            key={RandomId()}
          >
            {children}
          </Link>
        )
      )
    }
  </nav>
);

export default Navigation;
