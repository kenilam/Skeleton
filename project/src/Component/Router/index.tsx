import resources from '$/resources';
import {Transition} from '@/Component';
import React from 'react';
import {HashRouter as Provider, Redirect, Route, Switch, withRouter} from 'react-router-dom';
import IRouter from './spec';

const {view} = resources;

const Router: React.FunctionComponent<IRouter.Props> = (
  {
    appear,
    classNames,
    children,
    component,
    onEnter,
    onEntered,
    onEntering,
    onExit,
    onExiting,
    onExited,
    routeIndex,
    transitionIn,
    transitionStyle
  }
) => {
  const Component: React.FunctionComponent<IRouter.Component> = (
    {
      children, history: {location: {pathname}}, location
    }
  ) => {
    let enterFrame: number;
    let exitFrame: number;
    
    const routes = pathname === view.home.path
      ? view.home.name.toLowerCase()
      : pathname.substr(1).replace(/\//g, '.');
    
    const onEnterHandler: IRouter.OnEnter = (node, done) => {
      document.body.dataset.enteredRoutes = routes;
      
      window.cancelAnimationFrame(enterFrame);
      
      if (!onEnter) {
        return;
      }
      
      enterFrame = window.requestAnimationFrame(
        () => {
          onEnter(node, done);
        }
      );
    };
    
    const onExitHandler: IRouter.OnExit = node => {
      document.body.dataset.exitedRoutes = routes;
      
      window.cancelAnimationFrame(exitFrame);
      
      if (!onExit) {
        return;
      }
      
      exitFrame = window.requestAnimationFrame(
        () => {
          onExit(node);
        }
      );
    };
    
    return (
      <Transition
        appear={appear}
        classNames={classNames}
        component={component}
        onEnter={onEnterHandler}
        onEntered={onEntered}
        onEntering={onEntering}
        onExit={onExitHandler}
        onExiting={onExiting}
        onExited={onExited}
        transitionIn={transitionIn}
        transitionKey={location.pathname.split('/')[routeIndex + 1] || '/'}
        transitionStyle={transitionStyle}
      >
        <Switch location={location}>{children}</Switch>
      </Transition>
    );
  };
  
  const Instance = withRouter(Component);
  
  return (
    <Provider>
      <Instance>{children}</Instance>
    </Provider>
  );
};

export {Redirect, Route, Switch, Provider, withRouter};

export default Router;
