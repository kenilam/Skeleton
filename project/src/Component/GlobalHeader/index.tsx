import {CSSTransition, Logo, Navigation} from '@/Component';
import {TransitionStyleName} from '@/Component/CSSTransition';
import {Provider, withRouter} from '@/Component/Router';
import React from 'react';
import IGlobalHeader from './spec';
import Style from './Style';

const Component: React.FunctionComponent<IGlobalHeader.Component> = (
  {
    location,
    transitionInPaths
  }
) => (
  <CSSTransition
    transitionIn={
      transitionInPaths.some(
        pathname => pathname === location.pathname
      )
    }
    transitionStyle={TransitionStyleName.slideDown}
  >
    <header role='banner' data-component={Style.default}>
      <Logo />
      <Navigation inline={true} />
    </header>
  </CSSTransition>
);

const Instance = withRouter(Component);

const GlobalHeader: React.FunctionComponent<IGlobalHeader.Props> =
  ({transitionInPaths}) => (
    <Provider>
      <Instance transitionInPaths={transitionInPaths} />
    </Provider>
  );

export default GlobalHeader;
