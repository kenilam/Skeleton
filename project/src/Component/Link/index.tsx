import {Provider} from '@/Component/Router';
import React from 'react';
import {NavLink} from 'react-router-dom';
import ILink from './spec';
import Style from './Style';

const Link: React.FunctionComponent<ILink.Props> = (
  {
    children,
    className: wrapperClassName,
    component: Wrapper,
    onClick,
    to
  }) => {
  const Element = () => (
    <NavLink
      activeClassName={Style.active}
      data-component={Style.default}
      exact={true}
      onClick={onClick}
      to={to}
    >
      {children}
    </NavLink>
  );
  
  return (
    <Provider>
      {
        Wrapper ? (
          <Wrapper className={wrapperClassName}>
            <Element />
          </Wrapper>
        ) : <Element />
      }
    </Provider>
  );
};

export default Link;
