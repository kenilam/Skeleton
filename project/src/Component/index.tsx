import CSSTransition from './CSSTransition';
import './Font';
import Link from './Link';
import Logo from './Logo';
import Navigation from './Navigation';
import Router from './Router';
import Transition from './Transition';
import WebGL from './WebGL';

export {
  CSSTransition,
  Link,
  Logo,
  Navigation,
  Router,
  Transition,
  WebGL
};
