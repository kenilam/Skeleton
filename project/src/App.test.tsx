import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Component from './App';

describe('App', () => {
  test('should match with snapshot', () => {
    // @ts-ignore
    const renderer = new ShallowRenderer();
    expect(
      renderer.render(
        <Component />
      )
    ).toMatchSnapshot();
    
  });
});
