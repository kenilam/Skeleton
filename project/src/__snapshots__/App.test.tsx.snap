// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match with snapshot 1`] = `
<React.Fragment>
  <GlobalHeader
    transitionInPaths={
      Array [
        undefined,
        "/about",
        "/works",
        "/contact",
      ]
    }
  />
  <Router
    routeIndex={0}
    transitionStyle="fade"
  >
    <Route
      path="/about"
      render={[Function]}
    />
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      path="/works"
    />
    <Route
      component={[Function]}
    />
  </Router>
</React.Fragment>
`;
